_delb/xpath.py

```diff
@@ -51,19 +59,58 @@ def _split(expression: str, separator: str) -> Iterator[str]:
             quote = character
 
         part += character
+        cursor += 1
 
     yield part
 
 
-class XPathExpression:
+class XPathExpression(UserString):
+    __slots__ = ("location_paths",)
+
     def __init__(self, expression: str):
         self.location_paths = [LocationPath(x.strip()) for x in _split(expression, "|")]
 
-    def __str__(self):
+    @property
+    def data(self):
         return " | ".join(str(x) for x in self.location_paths)
 
+    def is_unambiguously_locatable(self) -> bool:
+        if len(self.location_paths) != 1:
+            return False
+
+        steps = self.location_paths[0].location_steps
+
+        if steps[0].axis != "self":
+            return False
+
+        for step in steps[1:]:
+            if step.axis != "child":
+                return False
+
+            predicates = step.predicates
+            if not predicates:
+                continue
+            if len(predicates) > 1:
```

Comment:
I don't get why you wouldn't allow this, even though it's implemented, and concatenating predicates with `and` is explicitly permitted..?

Reply:
i was and am very unsure that/how/which multiple predicates can be grouped in square brackets, b/c i find the [spec](https://www.w3.org/TR/1999/REC-xpath-19991116/#predicates) very unspecific regarding a definition and it gives only examples where one of the predicates tests for a node's index. the latter is ignored by the implementation b/c there's no use for it.

also, the data model is inconsistent with regards to predicates, it follows the simple parsing not the grammar, see [this](https://github.com/funkyfuture/delb/blob/main/_delb/xpath.py#L90) and [that](https://github.com/funkyfuture/delb/pull/20/commits/83f36cc9333db82cbcca3f9077b0dd65bc7ff1e5#diff-4cebee66d5fda6b56b40455efd0a93eba562dfc41c1b3f27ee8d678e29bff537R16).

Reply:
okay, definition `4` [here](https://www.w3.org/TR/1999/REC-xpath-19991116/#predicates#section-Location-Steps) clears that up for me.

Reply:
do you mean `Step ::= | AxisSpecifier NodeTest Predicate*`, and does that indicate you can concatenate as many individually-bracketed predicates as you want?

Reply:
yes, and yes that's how i'd interpret it. i couldn't find BNFs for / definitions of possible predicates yet, which is one reason why i'm hesistating to do it more properly.